### PR TITLE
[DEMO NO MERGE] Decouple procedure callers from mutator UI implemenation Mk2

### PR DESCRIPTION
**Pull request by BeksOmega**
## Basics
This is just a demo showing how the procedure callers could be generalized so that they are no longer tightly coupled to Blockly's current mutator-ui implimentation.

This is not for merging, and will be closed immediately after it is opened.

### Background
See #3724 

### Solution
This solution gives each parameter an ID that is separate from both the variable ID and the mutator block ID. It is very similar to the previous demo, but this one avoids coalescing vars accidentally.

This implementation brings back onFinishEditing. It's not necessarily necessary, you could handle deleting the previous var inside the validator. But I thought this was easier.

---
I think this is a pretty good for a demo, but if this was actually for PR there are still things I would want to clean up. Deleting arguments sometimes leaves unused vars in the flyout (although I think this happens on master). And I would want to explore refactoring `mutateCallers`.

### CODE REVIEW

**Comment by BeksOmega** [comment]
This is not for merging, it is just a demo.

File: blocks/procedures.js [hunk 1]
```diff
@@ -128,19 +127,27 @@ Blockly.Blocks['procedures_defnoreturn'] = {
...
         var varId = childNode.getAttribute('varid') || childNode.getAttribute('varId');
+        this.varIds_.push(varId);
         this.arguments_.push(varName);
         var variable = Blockly.Variables.getOrCreateVariablePackage(
             this.workspace, varId, varName, '');
         if (variable != null) {
           this.argumentVarModels_.push(variable);
         } else {
-          console.log('Failed to create a variable with name ' + varName + ', ignoring.');
+          console.log('Failed to create a variable with name ' +
```
Comment (rachel-fenichel) [inline blocks/procedures.js]: There's a pre-existing problem here, but it's worth addressing if you're changing this code: If no variable can be created we still add to most of the arrays (`arguments_`, `varIds_`, `paramIds_`) but not to `argumentVarModels_`.  That means they're out of sync permanently.

File: core/field.js [hunk 2]
```diff
@@ -852,7 +852,10 @@ Blockly.Field.prototype.setValue = function(newValue) {
     return;
   }
 
-  if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
+  if (this.sourceBlock_ &&
+      !this.sourceBlock_.disposed &&
```
Comment (rachel-fenichel) [inline core/field.js]: Is this a change we need no matter what?

Comment (BeksOmega) [reply core/field.js]: Ah, yep I forgot I wanted to do that. Pull request at #3766

**Comment by BeksOmega** [comment]
> Can I get a walkthrough on this? My understanding so far:

I think you've got it already haha

> * the goal is to change `setProcedureParameters` to not assume that the mutator UI has a stack of blocks

Correct : )

> * `paramId` is now separate from `varId`, but parallel

Yesh! This is a bit confusing because there's no concrete object that paramId is identifying. Really it's used in two cases:
1) We need to associate a new var with a position.
2) We need to change the positions of the vars.

If either one of these was static, we could use it as the ID instead. For example: If the position was static we could use that to associate new vars with each position. Or if the vars were static we could pass their IDs to the caller for reordering (which is what #3724 tried to do).

Because neither is static we need something else that is.

It would be nice if there was an clearer way to explain it though.

> * added `idsToConnections` for procedure caller blocks
> * `paramId` and `idsToConnections` are similar to the old `quarkIds` and `quarkConnections`

Yes and yes! All of this is so that when we handle case 2 the child blocks can get reconnected correctly.

File: blocks/procedures.js (hunk 1 above)
Comment (BeksOmega) [reply blocks/procedures.js]: Good idea!

**Comment by rachel-fenichel** [comment]
I'm interested in fixing this, but also really wary of PRs because we don't currently have tests. I just opened an issue (#3781) for procedure block tests, and I want to have tests in place before changing the actual block implementaiton.

**Pull request closed without merging**